File: pharmacy-api/public-api/broker_setup.go
package main

import (
	"log"

	shared_models "pharmacy-api/shared/models"
)

func setupBroker(broker *shared_models.Broker) (err error) {

	err = broker.Channel.ExchangeDeclare("public_exchange", "direct", true, false, false, false, nil)
	if err != nil {
		log.Fatal("failed to declare exchange:", err)
		return
	}

	queues := []struct {
		name       string
		routingKey string
	}{
		{"public_goods_queue", "goods"},
		{"public_schedule_queue", "schedule"},
		{"public_orders_queue", "orders"},
		{"public_announces_queue", "announces"},
	}

	for _, q := range queues {
		queue, err := broker.Channel.QueueDeclare(q.name, true, false, false, false, nil)
		if err != nil {
			log.Fatal("failed at declare queues:", err)
			return err
		}

		err = broker.Channel.QueueBind(queue.Name, q.routingKey, "public_exchange", false, nil)
		if err != nil {
			log.Fatal("failed at bindings queues:", err)
			return err
		}
	}

	log.Println("RabbitMQ setup completed successfully")
	return
}

File: pharmacy-api/src/public-api/broker_setup.ts
import { Channel, Options } from 'amqplib';
import { Broker } from '@pharmacy/src/shared/models/models';

export interface BrokerSetup {
  channel: Channel;
}

export async function setupBroker(broker: Broker): Promise<void> {
  try {
    const exchangeOptions: Options.AssertExchange = {
      durable: true,
      autoDelete: false,
      arguments: {}
    };
    
    await broker.channel.assertExchange("public_exchange", "direct", exchangeOptions);

    const queues = [
      { name: "public_goods_queue", routingKey: "goods" },
      { name: "public_schedule_queue", routingKey: "schedule" },
      { name: "public_orders_queue", routingKey: "orders" },
      { name: "public_announces_queue", routingKey: "announces" },
    ];

    for (const q of queues) {
      const queueOptions: Options.AssertQueue = {
        durable: true,
        autoDelete: false,
        exclusive: false,
      };
      
      const queue = await broker.channel.assertQueue(q.name, queueOptions);

      await broker.channel.bindQueue(queue.queue, "public_exchange", q.routingKey);
    }

    console.log("RabbitMQ setup completed successfully");
    console.log(`Exchange "public_exchange" declared with 4 queues bound`);
  } catch (err) {
    console.error("Failed to setup RabbitMQ:", err);
    throw err;
  }
}

export async function teardownBroker(broker: { channel: Channel }): Promise<void> {
  try {
    const queues = [
      "public_goods_queue",
      "public_schedule_queue",
      "public_orders_queue",
      "public_announces_queue",
    ];

    for (const queue of queues) {
      await broker.channel.deleteQueue(queue);
    }

    await broker.channel.deleteExchange("public_exchange");
    console.log("RabbitMQ teardown completed successfully");
  } catch (err) {
    console.error("Failed to teardown RabbitMQ:", err);
    throw err;
  }
}